jms/src/main/scala/akka/stream/alpakka/jms/JmsProducerStage.scala

```diff
@@ -100,41 +110,82 @@ private[jms] final class JmsProducerStage[A <: JmsMessage, PassThrough](settings
+        import settings.sendRetrySettings._
+        import send._
+        if (maxRetries < 0 || attempt + 1 < maxRetries) {
+          val nextAttempt = attempt + 1
+          val delay = waitTime(nextAttempt).min(maxBackoff)
+          scheduleOnce(send.copy(attempt = nextAttempt), delay)
+        } else {
+          holder(Failure(ex))
+        }
+      }
+
+      private val sendCompletedCB = getAsyncCallback[(SendAttempt[E], Try[Unit], JmsMessageProducer)] { tuple =>
+        val send = tuple._1
+        val outcome = tuple._2
+        val jmsProducer = tuple._3
+        // same epoch indicates that the producer belongs to the current alive connection.
+        if (jmsProducer.epoch == currentJmsProducerEpoch) jmsProducers.enqueue(jmsProducer)
+
+        import send._
+
+        outcome match {
+          case Success(_) =>
+            holder(Success(message))
+            pushNextIfPossible()
+          case Failure(t: JMSException) =>
+            nextTryOrFail(t, send)
+          case Failure(t) =>
+            holder(Failure(t))
+            handleFailure(t, holder)
+        }
+      }
+
       override def postStop(): Unit = {
         jmsSessions.foreach(_.closeSession())
         jmsConnection.foreach(_.close)
       }
 
-      private val futureCB = getAsyncCallback[Holder[E]](
-        holder =>
-          holder.elem match {
-            case Success(_) => pushNextIfPossible() // on success, try to push out the new element.
-            case Failure(ex) => handleFailure(ex, holder)
-        }
-      )
-
       private def pullIfNeeded(): Unit =
         if (jmsProducers.nonEmpty // only pull if a producer is available in the pool.
-            && !inFlightMessagesWithProducer.isFull // and a place is available in the in-flight queue.
+            && !inFlightMessages.isFull // and a place is available in the in-flight queue.
             && !hasBeenPulled(in))
           tryPull(in)
 
       private def pushNextIfPossible(): Unit =
-        if (inFlightMessagesWithProducer.isEmpty) {
+        if (inFlightMessages.isEmpty) {
           // no messages in flight, are we about to complete?
           if (isClosed(in)) completeStage() else pullIfNeeded()
-        } else if (inFlightMessagesWithProducer.peek().elem eq NotYetThere) {
+        } else if (inFlightMessages.peek().elem eq NotYetThere) {
           // next message to be produced is still not there, we need to wait.
           pullIfNeeded()
         } else if (isAvailable(out)) {
-          val holder = inFlightMessagesWithProducer.dequeue()
-          holder.jmsProducer.foreach(jmsProducers.enqueue) // put back jms producer to the pool.
+          val holder = inFlightMessages.dequeue()
           holder.elem match {
             case Success(elem) =>
               push(out, elem)
               pullIfNeeded() // Ask for the next element.
 
-            case Failure(NonFatal(ex)) => handleFailure(ex, holder)
+            case Failure(ex) => handleFailure(ex, holder)
```

Comment:
everything that is wrapped in a `Failure` is `NonFatal` (except you really try hard)